File: src/java/nz/co/searchwellington/feeds/FeedReaderRunner.java
package nz.co.searchwellington.feeds;

import java.util.List;
import java.util.Set;

import nz.co.searchwellington.mail.Notifier;
import nz.co.searchwellington.model.Feed;
import nz.co.searchwellington.model.LinkCheckerQueue;
import nz.co.searchwellington.repositories.ResourceRepository;

import org.apache.log4j.Logger;

public class FeedReaderRunner {

	Logger log = Logger.getLogger(FeedReaderRunner.class);
	
	private ResourceRepository resourceDAO;
	private FeedReader feedReader;
    private LinkCheckerQueue linkCheckerQueue;
    private Notifier notifier;

	    
	protected FeedReaderRunner(ResourceRepository resourceDAO, FeedReader feedReader, LinkCheckerQueue linkCheckerQueue, Notifier notifier) {
		this.resourceDAO = resourceDAO;
		this.feedReader = feedReader;
		this.linkCheckerQueue = linkCheckerQueue;
		this.notifier = notifier;
	}

	
	public void readSingleFeed(Feed feed) {
		 Set<Integer> acceptedNewsitemIds = this.feedReader.processFeed(feed.getId());
		 queueAcceptedNewsitemsIds(acceptedNewsitemIds);
	}
	
	
	public void readAllFeeds(List<Feed> feeds) {
		log.info("Running Reader.");
		for (Feed feed: feeds) {      
			 queueAcceptedNewsitemsIds(this.feedReader.processFeed(feed.getId()));
		}
		log.info("Finished reading feeds.");		
	}

	
	private void queueAcceptedNewsitemsIds(Set<Integer> acceptedNewsitemIds) {
		for (Integer id : acceptedNewsitemIds) {
			log.info("Queuing for link check: " + id);
			linkCheckerQueue.add(id);
			notifier.sendAcceptanceNotification("Accepted newsitem from feed", resourceDAO.loadResourceById(id));
		}
	}
	
}

File: src/main/scala/nz/co/searchwellington/feeds/FeedReaderRunner.scala
package nz.co.searchwellington.feeds

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.model.{Feed, User}
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.apache.log4j.Logger
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.task.TaskExecutor
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component

import scala.concurrent.{Await, ExecutionContext, Future}

@Component class FeedReaderRunner @Autowired()(feedReader: FeedReader, mongoRepository: MongoRepository,
                                               feedReaderTaskExecutor: TaskExecutor) // TODO named bean
  extends ReasonableWaits {

  private val log = Logger.getLogger(classOf[FeedReaderRunner])
  private val FEED_READER_PROFILE_NAME = "feedreader"

  implicit val executionContext = ExecutionContext.fromExecutor(feedReaderTaskExecutor)

  @Scheduled(cron = "0 */10 * * * *")
  def readFeeds {
    def readAllFeeds(feeds: Seq[Feed]): Future[Boolean] = {
      getFeedReaderUser.map { maybyFeedUser =>
        maybyFeedUser.map { feedReaderUser =>
          log.info("Reading " + feeds.size + " feeds as user " + feedReaderUser.name)
          feeds.foreach { feed =>
            try {
              Await.result(feedReader.processFeed(feed, feedReaderUser), TenSeconds)
            } catch {
              case e: Exception =>
                log.error("Error reading feed: " + feed, e)
            }
          }
          true

        }.getOrElse {
          log.warn("Feed reader could not run as no user was found with profile name: " + FEED_READER_PROFILE_NAME)
          false
        }
      }
    }

    log.info("Running feed reader.")
    mongoRepository.getAllFeeds().flatMap { feeds =>
      readAllFeeds(feeds)
    }.map { _ =>
      log.info("Finished reading feeds.")
    }
  }

  private def getFeedReaderUser: Future[Option[User]] = mongoRepository.getUserByProfilename(FEED_READER_PROFILE_NAME)

}
